File: rust/5.最长回文子串.rs
/*
 * @lc app=leetcode.cn id=5 lang=rust
 *
 * [5] 最长回文子串
 */

// @lc code=start
impl Solution {
    pub fn longest_palindrome(s: String) -> String {
        fn is_palindrome(s: &str) -> bool {
            let mut s_l = s[..s.len() / 2].chars();
            let mut s_r = s[(s.len() + 1) / 2..].chars().rev();
            while let (Some(l), Some(r)) = (s_l.next(), s_r.next()) {
                if l != r {
                    return false;
                }
            }
            true
        }

        let mut result = String::from("");
        for length in (0..(s.len() + 1)).rev() {
            for index in 0..(s.len() - length + 1) {
                let piece = &s[index..index + length];

                if is_palindrome(piece) {
                    result = String::from(piece);
                    return result;
                }
            }
        }

        result
    }
}
// @lc code=end

File: 5.最长回文子串.go
/*
 * @lc app=leetcode.cn id=5 lang=golang
 *
 * [5] 最长回文子串
 */
package leetcode

// @lc code=start
func longestPalindrome(s string) string {
	isPalindrome := func(x string) bool {
		i, j := 0, len(x)-1
		for i < j {
			if x[i] != x[j] {
				return false
			}
			i++
			j--
		}
		return true
	}
	n := len(s)
	for len := n; len > 0; len-- {
		for i := 0; i+len <= n; i++ {
			j := i + len
			x := s[i:j]
			if isPalindrome(x) {
				return x
			}
		}
	}
	return s[:1]
}

func longestPalindromedp(s string) string {
	var n = len(s)
	var dp = make([][]bool, n)
	for i := range dp {
		dp[i] = make([]bool, n)
	}
	for i := 0; i < n; i++ {
		dp[i][i] = true
	}
	var maxLen, start = 1, 0
	for len := 2; len <= n; len++ {
		for i := 0; i < n-len+1; i++ {
			j := i + len - 1
			if s[i] != s[j] {
				dp[i][j] = false
			} else {
				if len <= 2 {
					dp[i][j] = true
				} else {
					dp[i][j] = dp[i+1][j-1]
				}
			}
			if dp[i][j] && maxLen < len {
				maxLen = len
				start = i
			}
		}
	}
	return s[start : start+maxLen]
}

// @lc code=end
